### PR TITLE
Use identical source unit names in FileReader and CompilerStack

### LINKED ISSUE #11233: ICE in --ast-compact-json when --base-path affects an import path
## Description
With `--ast-compact-json` the compiler apparently has trouble finding files when `--base-path` is used and crashes:

``` bash
cd /tmp
echo 'import "contract2.sol";' > contract1.sol
touch contract2.sol

solc contract1.sol --ast-compact-json --base-path /tmp
```
```
Exception during output generation: /solidity/libsolidity/interface/CompilerStack.cpp(1416): Throw in function const solidity::frontend::CompilerStack::Source& solidity::frontend::CompilerStack::source(const string&) const
Dynamic exception type: boost::wrapexcept<solidity::langutil::CompilerError>
std::exception::what: Given source file not found.
[solidity::util::tag_comment*] = Given source file not found.
```

Interestingly, as can be seen above, the crash happens despite the fact that `contract2.sol` should resolve to the same file with and without `--base-path /tmp`, since `/tmp` is the current working directory.

Removing `--base-path` option from the command resolves the problem.

## Environment

- Compiler version: 0.8.3
- Operating system: Arch Linux

### PR DESCRIPTION
**Pull request by cameel**
Fixes #11233. See comments in that issue for my analysis of the problem.

### CODE REVIEW

File: libsolidity/interface/FileReader.cpp [hunk 1]
```diff
@@ -81,7 +80,7 @@ ReadCallback::Result FileReader::readFile(string const& _kind, string const& _pa
 
 		// NOTE: we ignore the FileNotFound exception as we manually check above
 		auto contents = readFileAsString(canonicalPath.string());
-		m_sourceCodes[path.generic_string()] = contents;
+		m_sourceCodes[_sourceUnitName] = contents;
```
Comment (cameel) [inline libsolidity/interface/FileReader.cpp:83]: This is the part that fixes the bug. The rest of the PR is just fluff like tests, docs, refactors, etc.

File: test/cmdlineTests.sh [hunk 2]
```diff
@@ -162,7 +162,7 @@ function test_solc_behaviour()
         sed -i.bak -e 's/\(^[ ]*auxdata: \)0x[0-9a-f]*$/\1<AUXDATA REMOVED>/' "$stdout_path"
         sed -i.bak -e 's/ Consider adding "pragma .*$//' "$stderr_path"
         sed -i.bak -e 's/\(Unimplemented feature error.* in \).*$/\1<FILENAME REMOVED>/' "$stderr_path"
-        sed -i.bak -e 's/"version": "[^"]*"/"version": "<VERSION REMOVED>"/' "$stdout_path"
+        sed -i.bak -e 's/"version":[ ]*"[^"]*"/"version": "<VERSION REMOVED>"/' "$stdout_path"
```
Comment (cameel) [inline test/cmdlineTests.sh:165]: This a fix for version not being stripped from output of `test/cmdlineTests/combined_json_with_base_path/`.

**Review by leonardoalt — DISMISSED** [review]
Needs rebase, but looks good

**Comment by cameel** [comment]
Rebased. The conflict was only in the changelog.

BTW, I think this may affect the paths LSP uses (either positively or negatively, depending on what exactly it does with them), so I'd really like to have a review from @christianparpart.

**Review by chriseth — APPROVED** [review]

File: libsolidity/interface/FileReader.cpp (hunk 1 above)
Comment (christianparpart) [reply libsolidity/interface/FileReader.cpp:83]: if you pass the `_sourceUnitName` as value parameter you then could move that in here. Would this make sense?

**Comment by christianparpart** [comment]
> Rebased. The conflict was only in the changelog.
> 
> BTW, I think this may affect the paths LSP uses (either positively or negatively, depending on what exactly it does with them), so I'd really like to have a review from @christianparpart.

I will have a look right now. thanks for pinging. :)

File: libsolidity/interface/FileReader.cpp (hunk 1 above)
Comment (cameel) [reply libsolidity/interface/FileReader.cpp:83]: Yeah, but this is a callback so I think it has to follow the original signature exactly. I.e. I can change argument name but not necessarily the type.

**Pull request merged**